File: C#/Easy/FloodFill.cs
public class Solution
{
    public int[][] FloodFill(int[][] image, int sr, int sc, int color)
    {
        int nRows = image.Length;
        int nCols = image[0].Length;
        int newColor = image[sr][sc];

        void dfs(int r, int c)
        {
            if (r < 0 || c < 0 || r >= nRows || c >= nCols) return;
            if (image[r][c] != newColor || image[r][c] == color) return;

            image[r][c] = color;
            dfs(r + 1, c);
            dfs(r - 1, c);
            dfs(r, c + 1);
            dfs(r, c - 1);
        }

        dfs(sr, sc);
        return image;
    }
}

/*

    DFS

    Time: O(m * n)
    Space: O(m * n)

    Where m is the number of rows and n is the number of columns of the image

    ---------------------------------------------------------------

    Note: An important edge case to consider is when image[sr][sc] == color (it can lead to stack overflow).

*/

File: Swift/Easy/FloodFill.swift
class Solution {
    func floodFill(_ image: [[Int]], _ sr: Int, _ sc: Int, _ color: Int) -> [[Int]] {
        if image[sr][sc] == color { return image }

        var image = image
        let nRows = image.count
        let nCols = image[0].count
        let oldColor = image[sr][sc]

        func performFloodFill(_ row: Int, _ col: Int) {
            if (row < 0 || col < 0 || row >= nRows || col >= nCols) { return }
            if (image[row][col] != oldColor) { return }
            
            image[row][col] = color
            performFloodFill(row + 1, col)
            performFloodFill(row - 1, col)
            performFloodFill(row, col + 1)
            performFloodFill(row, col - 1)
        }

        performFloodFill(sr, sc)
        return image
    }
}

/*

    DFS flood fill

    Time: O(m * n)
    Space: O(m * n)

*/